src/stim/util_top/reference_sample_tree.cc

```diff
@@ -0,0 +1,179 @@
+#include "stim/util_top/reference_sample_tree.h"
+
+using namespace stim;
+
+bool ReferenceSampleTree::empty() const {
+    if (repetitions == 0) {
+        return true;
+    }
+    if (!prefix_bits.empty()) {
+        return false;
+    }
+    for (const auto &child: suffix_children) {
+        if (!child.empty()) {
+            return false;
+        }
+    }
+    return true;
+}
+
+void ReferenceSampleTree::flatten_and_simplify_into(std::vector<ReferenceSampleTree> &out) const {
+    if (repetitions == 0) {
+        return;
+    }
+
+    std::vector<ReferenceSampleTree> flattened;
+    if (!prefix_bits.empty()) {
+        flattened.push_back(ReferenceSampleTree{
+            .prefix_bits = prefix_bits,
+            .suffix_children = {},
+            .repetitions = 1,
+        });
+    }
+    for (const auto &child : suffix_children) {
+        child.flatten_and_simplify_into(flattened);
+    }
+
+    // Fuse children.
+    auto &f = flattened;
+    for (size_t k = 0; k < f.size(); k++) {
+        // Combine children with identical contents by adding their rep counts.
+        while (k + 1 < f.size() && f[k].prefix_bits == f[k + 1].prefix_bits && f[k].suffix_children == f[k + 1].suffix_children) {
+            f[k].repetitions += f[k + 1].repetitions;
+            f.erase(f.begin() + k + 1);
+        }
+
+        // Fuse children with unrepeated contents if they can be fused.
+        while (k + 1 < f.size() && f[k].repetitions == 1 && f[k].suffix_children.empty() && f[k + 1].repetitions == 1) {
+            f[k].suffix_children = std::move(f[k + 1].suffix_children);
+            f[k].prefix_bits.insert(f[k].prefix_bits.end(), f[k + 1].prefix_bits.begin(), f[k + 1].prefix_bits.end());
+            f.erase(f.begin() + k + 1);
+        }
+    }
+
+    if (repetitions == 1) {
+        // Un-nest all the children.
+        for (auto &e : flattened) {
+            out.push_back(e);
+        }
+    } else if (flattened.size() == 1) {
+        // Merge with single child.
+        flattened[0].repetitions *= repetitions;
+        out.push_back(std::move(flattened[0]));
+    } else if (flattened.empty()) {
+        // Nothing to report.
+    } else if (flattened[0].suffix_children.empty() && flattened[0].repetitions == 1) {
```

Comment:
Pulling the leaf contents up into the interior node doesn't really work recursively, because it involves the interior node as well as the child. It can't happen entirely within the child. This complication is definitely a downside of defining the nodes as *both* a payload and children and trying to use both efficiently, instead of only one or the other.

Reply:
I think the structure (node = payload + children) is fine. I'm referring to the theoretical case that you have
`vector { node { prefix: XXX }, node { prefix: XXXXX }, node: { prefix: XXXX }`
or `node { children: node { children: node { children: node { prefix: XXXX } } } }`
etc (plus intermixing)
which would hit cases above recursively (and should be fully unwrapped).

Reply:
Ah. I see now. Such recursion shouldn't be necessary/possible because it's handled (collapsed) at each recursive call of `flatten_and_simplify_into()`.